hypothesis-python/src/hypothesis/extra/array_api.py

```diff
@@ -760,10 +829,13 @@ def indices(
     )
 
 
-def make_strategies_namespace(xp: Any) -> SimpleNamespace:
+def make_strategies_namespace(
+    xp: Any, *, api_version: Union["Ellipsis", NominalVersion] = ...
+) -> SimpleNamespace:
```

Comment:
We should probably add some caching with a `WeakValuesDict` to help mitigate startup latency issues.

Reply:
TIL about the `weakref` module. So something like this?

```python
args_to_xps = {}

def make_strategies_namespace(xp, api_version):
    try:
        namespace = args_to_xps[(xp, api_version)]
    except (KeyError, TypeError):
        pass
    else:
        return namespace
    ...
    namespace = PrettySimpleNamespace(**kwargs)
    try:
        args_to_xps[(xp, api_version)] = namespace
    except TypeError:
        pass
    return namespace
```

Rudimentary test that does pass with this, just not sure if there's any other concerns I'm missing.

```python
def test_make_strategies_namespace(xp, xps, monkeypatch):
    try:
        hash(xp)
    except TypeError:
        pytest.skip("TODO")
    monkeypatch.setattr(array_api, "args_to_xps", WeakValueDictionary())
    assert len(array_api.args_to_xps) == 0  # sanity check
    xps1 = array_api.make_strategies_namespace(xp, api_version="2021.12")
    assert len(array_api.args_to_xps) == 1
    xps2 = array_api.make_strategies_namespace(xp, api_version="2021.12")
    assert len(array_api.args_to_xps) == 1
    assert xps2 is xps1
    del xps1
    del xps2
    assert len(array_api.args_to_xps) == 0
```

(probably I'll want to support parametrizing just over `xp` and not `xps` generally in the test suite, and sanity check it supports "2021.12" here. or something fancier, will have a think.)

Reply:
That's it! Just add a leading underscore to the name, and use a weakvaluesdict instead of a regular dict 😁